Adds support for georeferenced coordinates.

arjo129 opened the pull request:
## New feature implementation

### Implemented feature

Adds support for visualizing OSM maps and using Geographic coordinates (Lat/Lon) instead of just (x, y) coordinates.

**To use**:
Go to `tools`->`Georeference`->`Set Reference...`. Enter latitude and longitude of reference. Press `Set Reference`. You should see the maps loading.
You may also `set reference via anchor` or move an anchor to a specified latitude/longitude using `Move To Lat/Lon`
You can adjust the map resolution using the bottom slider.
To view an offset you can use the `Tools`->`Georeference`->`View Reference as UTM`.
![Screenshot from 2023-06-15 13-04-39](https://github.com/open-rmf/rmf_site/assets/542272/afd2d345-abc3-4740-af1c-9aa87ebc7205)
## Known Issues
* We really should switch to Proj, but for now we use the `utm` crate because `Proj`'s rust binding does not yet support WASM
* Map tiles should ideally be drawn in web mercator frame, however this would require us to warp the units of our building graph to support web mercator and then have those warped back to `UTM`.
* Changing resolution causes tiles to despawn and respawn. It would be better if the effect usses a fade in-fade out animation.
* Currently we use OSM's main service. Ideally we should be able to switch tile providers in the future.

luca-della-vedova commented:
A few usability feedback points before diving too deep into the code, from trying to give it a spin:

It seems the `Set Reference via Anchor` menu button doesn't work if it's clicked on a fresh project, it only works after a reference has already been set. I'd suggest either making sure it works on a fresh project or greying it out (I suppose the first since it's meant to be a way to set a reference?)

The anchor reference, view reference / tile resolution windows cannot be closed. The last can maybe be kept but the first two sound a bit unnecessary to always keep?

I tried the `move to lat lon` with no anchor selected and got a panic:
```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', rmf_site_editor/src/site/georeference.rs:135:33
```

If the `Move to Lat/Lon` button is clicked when the `Set Reference` window is open there will be glitches in egui, I believe due it might be due to reuse of widget names.

I tried to set the georeference of the office world and there are a few glitches between tiles, like black gaps and misalignments
![image](https://github.com/open-rmf/rmf_site/assets/9111558/ef3bb929-0d6c-4f98-9a77-2421d85e3ca4)

When I tried to use `Set reference via anchor` in a world where a reference already exists, all the tiles disappeared and I got a bunch of warnings:

```
2023-06-16T03:22:07.336874Z  WARN bevy_ecs::world: error[B0003]: Could not despawn entity 878v0 because it doesn't exist in this World.
2023-06-16T03:22:07.336898Z  WARN bevy_ecs::world: error[B0003]: Could not despawn entity 879v0 because it doesn't exist in this World.
2023-06-16T03:22:07.336903Z  WARN bevy_ecs::world: error[B0003]: Could not despawn entity 880v0 because it doesn't exist in this World.
2023-06-16T03:22:07.336906Z  WARN bevy_ecs::world: error[B0003]: Could not despawn entity 881v0 because it doesn't exist in this World.
``` 

For nice to haves:

I wonder if we should have some sort of visualization for the reference anchor when it's used as a georeference, maybe adding an entity attached to it similar to how `Location`s work? It's basically impossible right now to find the anchor that was set as a reference, since all  that can be seen is its entity ID and clicking the button with its entity ID enters the "select anchor" mode.

It would be great if we can add parsing of legacy building projects that are in wgs84, like the rmf_demos `campus` world which currently fails importing:
```
Failed loading legacy building Message("unknown variant `wgs84`, expected `reference_image` or `cartesian_meters`", Some(Pos { marker: Marker { index: 19, line: 1, col: 19 }, path: "coordinate_system" }))
```

Imho it would also be nice if we can update the embedded demo map under `assets/` to be georeferenced with the actual office location, to really showcase this feature

arjo129 commented:
Thanks for breaking it... I've tried to break the issues down into a working list.
> I tried the move to lat lon with no anchor selected and got a panic:
Fixed in [944fdd5](https://github.com/open-rmf/rmf_site/pull/133/commits/944fdd56066c336bd6dad81ed39c0639fee8689c)

>If the Move to Lat/Lon button is clicked when the Set Reference window is open there will be glitches in egui, I believe due it might be due to reuse of widget names.
Fixed in [944fdd5](https://github.com/open-rmf/rmf_site/pull/133/commits/944fdd56066c336bd6dad81ed39c0639fee8689c)

> I tried to set the georeference of the office world and there are a few glitches between tiles, like black gaps and misalignments
This an issue of web_mercator vs UTM. Its a whole can of worms that will need to be addressed. Our buildings are currently in meter units. However OSM maps are in "web mercator" pixels... There will need to be some deformation between the two. Ideally the tile sizes should also be in UTM.

>When I tried to use Set reference via anchor in a world where a reference already exists, all the tiles disappeared and I got a bunch of warnings:
Currently investigating.

> It would be great if we can add parsing of legacy building projects that are in wgs84, like the rmf_demos campus world which currently fails importing:
That's a separate PR. Hopefully will be less buggy.

> Imho it would also be nice if we can update the embedded demo map under assets/ to be georeferenced with the actual office location, to really showcase this feature
Agreed.

xiyuoh commented:
> When I tried to use Set reference via anchor in a world where a reference already exists, all the tiles disappeared and I got a bunch of warnings

For me if i re-check the `View Map` box, the tiles will reappear with the correct location

Also agreed on not being able to close widgets other than Set Geographic Reference. Maybe it would help if they are all grouped under a single widget too? Found myself having to move things around a little

arjo129 commented:
I've reworked the implementation significantly for this iteration and significantly simplified the user experience. It is probably ready for another round of review. The gaps between tiles have been eliminated. There are several ways to use this feature.

1. By loading an existing world.

You may download an existing world [here](https://github.com/open-rmf/rmf_site/files/12414536/roro_world2.site.zip) and open it. You should be able to see the world in the screen like so (notice no more gaps):

![image](https://github.com/open-rmf/rmf_site/assets/542272/758ba2c0-5824-442b-8bd0-6fd848d58e68)

You should also be able to toggle visibility from the "view" checklist. To enable this I added an API that allows us to extend menus via the ECS like what was done in #163

[Screencast from 2023-08-23 10-35-10.webm](https://github.com/open-rmf/rmf_site/assets/542272/6e176eed-b7d8-4ee1-8442-1a60b97aa26e)

You may also view the reference from the tools menu and change the tile reference via the settings. Note: There is a hard-coded tile limit setting the code (this is to prevent the cache size from exploding and us from hammering the OSM server). We should make this configurable in a future PR.
[Screencast from 2023-08-23 10-38-32.webm](https://github.com/open-rmf/rmf_site/assets/542272/e2b8d218-3f34-42e6-af1b-21123a76f52a)

2. From a world without any existing reference.

We can use the example world. Open the `Tool` > `Set reference` panel. This should allow you to set an origin latitude and longitude. 
![image](https://github.com/open-rmf/rmf_site/assets/542272/e15e99ab-8ff7-44a0-8ee9-f481c7121cf7)

You may also use the Inspect Anchor tool to move anchors to desired latitudes and longitudes.
![image](https://github.com/open-rmf/rmf_site/assets/542272/42fad156-22d3-43f6-bb50-c111dd407127)

We don't yet support using an anchor as a reference (that'll be a follow up PR) and we don't support lat/lon in theinspect Lane dialog (Also should be a follow up dialog).

mxgrey commented on rmf_site_format/src/asset_source.rs:
```diff
@@ -40,6 +41,7 @@ impl AssetSource {
             Self::Search(_) => "Search",
             Self::Bundled(_) => "Bundled",
             Self::Package(_) => "Package",
+            Self::OSMSlippyMap(_, _, _) => "Map",
```
Consider changing this to something like OSM Tile

arjo129 replied:
Worth noting that the protocol is called "slippy tiles"

[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

mxgrey commented on rmf_site_format/src/asset_source.rs:
```diff
@@ -30,6 +30,7 @@ pub enum AssetSource {
     Search(String),
     Bundled(String),
     Package(String),
+    OSMSlippyMap(i32, f32, f32),
```
What do you think of making this a struct variant instead of a tuple variant, e.g.
```
OSMSlippyMap{ zoom: i32, latitude: f32, longitude: f32 }
```
Otherwise I think it's unclear what each of the fields in the tuple is for.

arjo129 replied:
[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

mxgrey reviewed (changes requested):
This new feature is very awesome! I've left feedback on some minor details, but fundamentally these changes are excellent.

arjo129 commented on rmf_site_editor/src/site/georeference.rs:
```diff
@@ -0,0 +1,609 @@
+pub struct GeoReferenceSelectAnchorEvent {}
+
+#[derive(Debug, Clone)]
+pub struct GeoReferenceSetReferenceEvent;
```
Ah forgot to delete these.
[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

arjo129 commented on rmf_site_editor/src/osm_slippy_map.rs:
```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (C) 2023 Intrinsic LLC
```
[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

arjo129 commented on rmf_site_editor/src/osm_slippy_map.rs:
```diff
@@ -0,0 +1,334 @@
+    pub fn get_quad_mesh(&self) -> Option<Mesh> {
+        let nw = self.get_nw_corner();
+        let Ok(nw) = self.get_transform_from_lat_lon(nw.0, nw.1) else {
+            println!("Failed to get nw {:?}", nw);
```
[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

arjo129 commented on rmf_site_editor/src/site/georeference.rs:
```diff
@@ -0,0 +1,609 @@
+const MAX_TILES: usize = 50;
+
+#[derive(Debug, Clone)]
+pub struct GeoReferenceSelectAnchorEvent {}
```
[fc2c182](https://github.com/open-rmf/rmf_site/pull/133/commits/fc2c182d5679b8cf20298741363d855b0c0b2e5d)

mxgrey reviewed (approved):
I tweaked the syntax here to make it a bit nicer: [b7b452e](https://github.com/open-rmf/rmf_site/pull/133/commits/b7b452ee912389896615d5bce64405346588dc4b)

Thanks for this feature, this is awesome!
